agora/Agora/Governor/Scripts.hs

```diff
@@ -0,0 +1,746 @@
+                      pconstant ()
+                  )
+
+            gatOutputValidator = gatOutputValidator' # effectGroup
+
+        popaque $
+          pfoldr
+            # gatOutputValidator
+            # pconstant ()
+            # outputsWithGAT
+      PMutateGovernor _ -> P.do
+        popaque $ singleAuthorityTokenBurned patSymbol ctx.txInfo txInfo.mint
+  where
+    pgtValueOf :: Term s (PValue :--> PDiscrete GTTag)
+    pgtValueOf = phoistAcyclic $ pvalueDiscrete' gov.gtClassRef
+
+    patSymbol :: Term s PCurrencySymbol
+    patSymbol = phoistAcyclic $ pconstant $ atSymbolFromGovernor gov
+
+    ppstSymbol :: Term s PCurrencySymbol
+    ppstSymbol =
+      let AssetClass (sym, _) = proposalSTAssetClassFromGovernor gov
+       in phoistAcyclic $ pconstant sym
+
+    proposalDatumValid' :: Term s (PProposalDatum :--> PBool)
+    proposalDatumValid' =
+      let params = proposalFromGovernor gov
+       in phoistAcyclic $ proposalDatumValid params
+
+    pproposalValidatorAddress :: Term s PAddress
+    pproposalValidatorAddress =
+      let vh = proposalValidatorHashFromGovernor gov
+       in phoistAcyclic $ pconstant $ validatorHashToAddress vh
+
+    pstakeValidatorAddress :: Term s PAddress
+    pstakeValidatorAddress =
+      let vh = stakeValidatorHashFromGovernor gov
+       in phoistAcyclic $ pconstant $ validatorHashToAddress vh
+
+    psstSymbol :: Term s PCurrencySymbol
+    psstSymbol =
+      let sym = stakeSTSymbolFromGovernor gov
+       in phoistAcyclic $ pconstant sym
+
+    pgstSymbol :: Term s PCurrencySymbol
+    pgstSymbol =
+      let sym = governorSTSymbolFromGovernor gov
+       in phoistAcyclic $ pconstant sym
+
+--------------------------------------------------------------------------------
+
+governorSTSymbolFromGovernor :: Governor -> CurrencySymbol
+governorSTSymbolFromGovernor gov = mintingPolicySymbol policy
+  where
+    policy :: MintingPolicy
+    policy = mkMintingPolicy $ governorPolicy gov
+
+{- | Get the 'AssetClass' of GST from 'Governor'.
+ TODO: tag GST?
+-}
+governorSTAssetClassFromGovernor :: Governor -> AssetClass
+governorSTAssetClassFromGovernor gov = AssetClass (symbol, "")
+  where
+    symbol :: CurrencySymbol
+    symbol = governorSTSymbolFromGovernor gov
+
+-- | Get the `CurrencySymbol` of GAT from 'Governor'.
+atSymbolFromGovernor :: Governor -> CurrencySymbol
+atSymbolFromGovernor gov = mintingPolicySymbol policy
+  where
+    at = AuthorityToken $ governorSTAssetClassFromGovernor gov
+    policy = mkMintingPolicy $ authorityTokenPolicy at
+
+proposalSTAssetClassFromGovernor :: Governor -> AssetClass
+proposalSTAssetClassFromGovernor gov = AssetClass (symbol, "")
+  where
+    gstAC = governorSTAssetClassFromGovernor gov
+    -- JUSTIFICATIONL: the PST policy doesn't care about the following two fields at all.
+    -- FIXME: refactor PST policy, parameterize it only with GST assetclass or something.
```

Comment:
This can be a problem. The thing is, we can't provide the assetclass of the stake state token here, cause it will be tagged with the address of the stake validator, however stake validator requires the `Stake` parameter, which in turn requires the assetclass of proposal state token.

Reply:
cc @emiflake

Reply:
Okay, so the problem is that the `proposalPolicy` shouldn't take `Proposal` at all. It should take only the governor asset class. As shown by the diagram in my previous comment: https://github.com/Liqwid-Labs/agora/pull/61#issuecomment-1102342152

Reply:
Great, I'll fix that today.

Reply:
Thanks!